File: python/open_guardrail/guards/consent_language.py
"""Validate consent language quality and detect dark patterns."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_PATTERNS: list[tuple[str, re.Pattern[str]]] = [
    (
        "pre_ticked_consent",
        re.compile(
            r"pre[- ]?(ticked|checked|selected)"
            r"\s+(box|checkbox|consent)",
            re.IGNORECASE,
        ),
    ),
    (
        "bundled_consent",
        re.compile(
            r"by\s+(signing|agreeing|accepting)"
            r"\s+you\s+(also|additionally)\s+consent",
            re.IGNORECASE,
        ),
    ),
    (
        "hidden_terms",
        re.compile(
            r"hidden\s+(terms|conditions|clauses)",
            re.IGNORECASE,
        ),
    ),
    (
        "by_continuing_agree",
        re.compile(
            r"by\s+(continuing|using|accessing"
            r"|proceeding)\s+(you\s+)?"
            r"(agree|consent|accept)",
            re.IGNORECASE,
        ),
    ),
    (
        "dark_pattern_urgency",
        re.compile(
            r"(act\s+now|limited\s+time"
            r"|don'?t\s+miss|hurry)",
            re.IGNORECASE,
        ),
    ),
    (
        "missing_opt_out",
        re.compile(
            r"no\s+(option|way|ability)\s+to"
            r"\s+(opt[- ]?out|unsubscribe|withdraw)",
            re.IGNORECASE,
        ),
    ),
    (
        "forced_consent",
        re.compile(
            r"(must|required\s+to)"
            r"\s+(agree|consent|accept)"
            r"\s+to\s+(all|every)",
            re.IGNORECASE,
        ),
    ),
    (
        "deceptive_button",
        re.compile(
            r"(accept\s+all|agree\s+to\s+all"
            r"|yes\s+to\s+everything)",
            re.IGNORECASE,
        ),
    ),
    (
        "auto_enroll",
        re.compile(
            r"auto(matically)?\s*(enroll|subscribe"
            r"|opt[- ]?in)",
            re.IGNORECASE,
        ),
    ),
]


class _ConsentLanguage:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "consent-language"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for name, pat in _PATTERNS:
            if pat.search(text):
                matched.append(name)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0) if triggered else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="consent-language",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Consent language issue detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched": matched} if triggered else None
            ),
        )


def consent_language(
    *, action: str = "block"
) -> _ConsentLanguage:
    return _ConsentLanguage(action=action)

File: packages/guards/src/consent-language.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ConsentLanguageOptions {
  action: 'block' | 'warn';
}

const PATTERNS: { name: string; re: RegExp }[] = [
  { name: 'pre_ticked_consent', re: /pre[- ]?(ticked|checked|selected)\s+(box|checkbox|consent)/i },
  { name: 'bundled_consent', re: /by\s+(signing|agreeing|accepting)\s+you\s+(also|additionally)\s+consent/i },
  { name: 'hidden_terms', re: /hidden\s+(terms|conditions|clauses)/i },
  { name: 'by_continuing_agree', re: /by\s+(continuing|using|accessing|proceeding)\s+(you\s+)?(agree|consent|accept)/i },
  { name: 'dark_pattern_urgency', re: /(act\s+now|limited\s+time|don'?t\s+miss|hurry)/i },
  { name: 'missing_opt_out', re: /no\s+(option|way|ability)\s+to\s+(opt[- ]?out|unsubscribe|withdraw)/i },
  { name: 'forced_consent', re: /(must|required\s+to)\s+(agree|consent|accept)\s+to\s+(all|every)/i },
  { name: 'deceptive_button', re: /(accept\s+all|agree\s+to\s+all|yes\s+to\s+everything)/i },
  { name: 'buried_opt_out', re: /opt[- ]?out\s+(is\s+)?(buried|hidden|difficult|hard\s+to\s+find)/i },
  { name: 'auto_enroll', re: /auto(matically)?\s*(enroll|subscribe|opt[- ]?in)/i },
];

export function consentLanguage(options: ConsentLanguageOptions): Guard {
  return {
    name: 'consent-language',
    version: '0.1.0',
    description: 'Validates consent language quality and detects dark patterns',
    category: 'compliance',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const p of PATTERNS) {
        if (p.re.test(text)) {
          matched.push(p.name);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'consent-language',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched } : undefined,
      };
    },
  };
}
